### PR TITLE
support configuring component log levels

### PR DESCRIPTION
**Pull request by snowp**
Adds support for configuring component log levels via the EngineBuilder interfaces, allowing for more granular control over
what logs are emitted.

Risk Level: Low, new functionality
Testing: Android and iOS integration tests
Docs Changes: inline
Release Notes: n/a
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

File: library/common/http/client.h [hunk 1]
```diff
@@ -228,9 +228,7 @@ class Client : public Logger::Loggable<Logger::Id::http> {
     void readDisable(bool disable) override;
     uint32_t bufferLimit() override { return 65000; }
     // Not applicable
-    void setAccount(Buffer::BufferMemoryAccountSharedPtr) override {
-      PANIC("buffer accounts unsupported");
```
Comment (jingwei99) [inline library/common/http/client.h:232]: what's the reason/thoughts behind this?

(this is nothing critical, but just want to know)

File: library/common/engine.h [hunk 2]
```diff
@@ -31,12 +31,18 @@ class Engine : public Logger::Loggable<Logger::Id::main> {
    */
   ~Engine();
 
+  struct EngineRunOptions {
```
Comment (jingwei99) [inline library/common/engine.h:34]: 👍

File: library/common/http/client.h (hunk 1 above)
Comment (jingwei99) [reply library/common/http/client.h:232]: 👍

File: test/common/engine_test.cc [hunk 3]
```diff
@@ -79,12 +80,58 @@ TEST_F(EngineTest, EarlyExit) {
...
 
   engine_.reset();
   ASSERT_TRUE(test_context.on_exit.WaitForNotificationWithTimeout(absl::Seconds(3)));
 
   start_stream(0, {}, false);
 }
+
+TEST_F(EngineTest, ComponentLogLevel) {
+  const std::string level = "debug";
+  const std::string component_log_level = "assert:info,misc:trace";
```
Comment (jingwei99) [inline test/common/engine_test.cc:94]: the intent of this test as I assume is to verify engine will be running ok when the `component_log_level` is set correctly (or in the format that's expected), is this assumption right?

**Review by jingwei99 — DISMISSED** [review]
looks good to me

File: test/common/engine_test.cc (hunk 3 above)
Comment (snowp) [reply test/common/engine_test.cc:94]: I wrote this originally to verify that plumbing through the component log level without crashing (so yes your assumption is correct), though this test might be unnecessary now that we there are integration tests for Kotlin and iOS. The other test in here which validates that we don't crash when we get an invalid string is more valid.

Comment (jingwei99) [reply test/common/engine_test.cc:94]: 👌

File: library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt [hunk 4]
```diff
@@ -26,6 +26,7 @@ open class EngineBuilder(
     EnvoyEngineImpl(onEngineRunning, logger, eventTracker)
   }
   private var logLevel = LogLevel.INFO
+  private var componentLogLevel = ""
```
Comment (buildbreaker) [inline library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:29]: What are the options for component log level? If it's a finite set of strings, would making an enum configuration make more sense?

Comment (snowp) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:29]: It's a string of the format `<component>:<level>,<comopnent>:<level>`, and the components needs to be statically registered in Envoy. I started this with a map, but was deterred by the fact that passing a map through the C layer is painful, and decomposing the map into the string in the platform layer felt repetitive.

Open to suggestions though :)

Comment (rebello95) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:29]: The current form does feel a little magical. If we don't want to do a map of `component (string): LogLevel (enum)`, let's at least document the expected structure inline and in the documentation for the builder

Comment (buildbreaker) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:29]: I'm fine with the proposed changes but I would like to understand where these values are registered in Envoy. 

We can fixate those values on the platform layer which would make the validation in the common layer more sane while providing a clear set of values for the API.

Comment (snowp) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:29]: The components are statically registered here: https://github.com/envoyproxy/envoy/blob/main/source/common/common/logger.h#L29-L79

If we wanted to we could add a C API that let us query https://github.com/envoyproxy/envoy/blob/main/source/common/common/logger.h#L312 in order to get the list of registered log names. Another option would be to just capture a static list of valid log components that we manually keep in sync.

LMK which option ya'll prefer between the above and just documenting the expected log component string, happy to do whichever

Comment (snowp) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:29]: @buildbreaker ping on this, lmk how you want to proceed

File: library/common/engine.h (hunk 2 above)
Comment (goaway) [reply library/common/engine.h:34]: I like this conceptually, but I do think perhaps config should be separate from it. There's probably more that can go in here down the road.

File: library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt (hunk 4 above)
Comment (goaway) [reply library/kotlin/io/envoyproxy/envoymobile/EngineBuilder.kt:29]: I think posting the above link to the components and having an example string in the public API comments/docstrings would be helpful.

**Pull request closed without merging**